libs/libcommon/include/common/logger_useful.h

```diff
@@ -3,16 +3,83 @@
 /// Macros for convenient usage of Poco logger.
 
 #include <sstream>
+#include <atomic>
 #include <Poco/Logger.h>
+#include <Poco/Message.h>
+#include <Poco/Version.h>
+#include <Core/SettingsCommon.h>
+#include <Common/CurrentThread.h>
 
 #ifndef QUERY_PREVIEW_LENGTH
 #define QUERY_PREVIEW_LENGTH 160
 #endif
 
 using Poco::Logger;
+using Poco::Message;
+using DB::LogsLevel;
+using DB::CurrentThread;
 
 /// Logs a message to a specified logger with that level.
 
+#if defined(POCO_CLICKHOUSE_PATCH)
+
+#define LOG_TRACE(logger, message) do { \
+    const bool is_clients_log = (CurrentThread::getGroup() != nullptr) && (CurrentThread::getGroup()->client_logs_level >= LogsLevel::trace); \
+    if ((logger)->trace() || is_clients_log) {\
+    std::stringstream oss_internal_rare; \
+    oss_internal_rare << message; \
+    if (is_clients_log) {\
+        (logger)->force_log(oss_internal_rare.str(), Message::PRIO_TRACE); \
+    } else { \
+        (logger)->trace(oss_internal_rare.str()); \
+    }}} while(false)
+
+#define LOG_DEBUG(logger, message) do { \
+    const bool is_clients_log = (CurrentThread::getGroup() != nullptr) && (CurrentThread::getGroup()->client_logs_level >= LogsLevel::debug); \
+    if ((logger)->debug() || is_clients_log) {\
+    std::stringstream oss_internal_rare; \
+    oss_internal_rare << message; \
+    if (is_clients_log) {\
+        (logger)->force_log(oss_internal_rare.str(), Message::PRIO_DEBUG); \
+    } else { \
+        (logger)->debug(oss_internal_rare.str()); \
+    }}} while(false)
+
+#define LOG_INFO(logger, message) do { \
+    const bool is_clients_log = (CurrentThread::getGroup() != nullptr) && (CurrentThread::getGroup()->client_logs_level >= LogsLevel::information); \
+    if ((logger)->information() || is_clients_log) {\
+    std::stringstream oss_internal_rare; \
+    oss_internal_rare << message; \
+    if (is_clients_log) {\
```

Comment:
Why we need this check?

Reply:
Because we don't want to print the "trace" log if the client's level is "none" and server's level is "information", for example. force_log function won't check the level

Reply:
But you've already checked it three lines above.

Reply:
Ohhh it's copypaste